kernel/rros/observable.rs

```diff
@@ -0,0 +1,970 @@
+            poll_head: RrosPollHead::new(),
+            wake_irqwork: IrqWork::new(),
+            flush_irqwork: IrqWork::new(),
+            lock: HardSpinlock::new(),
+            serial_counter: 0,
+            writable_observers: 0,
+        }
+    }
+}
+
+#[repr(C)]
+pub struct __RrosNotification {
+    pub tag: u32,
+    pub serial: u32,
+    pub issuer: i32,
+    pub event: ObservableValue,
+    pub date: ktime::Timespec64,
+}
+
+pub struct RrosNotificationRecord {
+    pub tag: u32,
+    pub serial: u32,
+    pub issuer: i32,
+    pub event: ObservableValue,
+    pub date: i64,
+    pub next: Links<RrosNotificationRecord>,
+}
+
+impl RrosNotificationRecord {
+    pub fn new() -> Self {
+        Self {
+            tag: 0,
+            serial: 0,
+            issuer: 0,
+            event: ObservableValue::new(),
+            date: 0,
+            next: Links::default(),
+        }
+    }
+}
+
+impl GetLinks for RrosNotificationRecord {
+    type EntryType = RrosNotificationRecord;
+    fn get_links(data: &Self::EntryType) -> &Links<Self::EntryType> {
+        &data.next
+    }
+}
+
+pub struct RrosObserver {
+    pub backlog_size: usize,
+    pub free_list: List<Box<RrosNotificationRecord>>,
+    pub pending_list: List<Box<RrosNotificationRecord>>,
+    pub next: Links<RrosObserver>,
+    pub fundle: u32,
+    pub flags: i32,
+    pub refs: i32,
+    pub last_notice: RrosNotice,
+    // TODO: how to implement a zero-sized array, and is it essential?
+    // If considering the purpose of caching, I think either don't use backlog, or make backlog also become List<Box<RrosNotificationRecord>>
+    // backlog: List<Box<RrosNotificationRecord>>,
+}
+
+impl RrosObserver {
+    pub fn new() -> Self {
+        Self {
+            backlog_size: 0,
+            free_list: List::default(),
+            pending_list: List::default(),
+            next: Links::default(),
+            fundle: 0,
+            flags: 0,
+            refs: 0,
+            last_notice: RrosNotice::new(),
+            // backlog: List::new(),
+        }
+    }
+}
+
+impl GetLinks for RrosObserver {
+    type EntryType = RrosObserver;
+    fn get_links(data: &Self::EntryType) -> &Links<Self::EntryType> {
+        &data.next
+    }
+}
+
+fn add_subscription(
+    observable: &mut RrosObservable,
+    backlog_count: u32,
+    op_flags: i32,
+) -> Result<i32> {
+    let mut sbr = dovetail::dovetail_current_state().subscriber() as *mut RrosSubscriber;
+    if backlog_count == 0 {
+        return Err(kernel::Error::EINVAL);
+    }
+    if op_flags & (!RROS_NOTIFY_MASK) != 0 {
+        return Err(kernel::Error::EINVAL);
+    }
+    if sbr == 0 as *mut RrosSubscriber {
+        match Box::try_new(RrosSubscriber::new()) {
+            Ok(s) => sbr = Box::into_raw(s),
+            Err(_) => return Err(kernel::Error::ENOMEM),
+        }
+
+        let mut box_sbr = unsafe { Box::from_raw(sbr) };
+        box_sbr.as_mut().init();
+        sbr = Box::into_raw(box_sbr);
```

Comment:
The `new` and `init` workflow is twisty. Just delete `into_raw` in line 237 and `from_raw` in line 241.

Reply:
I will delete them.